src/agent/src/image_rpc.rs

```diff
@@ -0,0 +1,160 @@
+
+        let mut pull_command = Command::new(SKOPEO_PATH);
+        pull_command
+            .arg("--insecure-policy")
+            .arg("copy")
+            .arg(source_image)
+            .arg(&target_path_manifest);
+
+        if let Some(source_creds) = source_creds {
+            pull_command.arg("--src-creds").arg(source_creds);
+        }
+
+        let status: ExitStatus = pull_command.status()?;
+
+        if !status.success() {
+            return Err(anyhow!(format!("failed to pull image: {:?}", status)));
+        }
+
+        // Copy image from one local file-system to another
+        // Resulting image is still stored in manifest format, but no longer includes the signature
+        // The image with a signature can then be unpacked into a bundle
+        let status: ExitStatus = Command::new(SKOPEO_PATH)
+            .arg("--insecure-policy")
+            .arg("copy")
+            .arg(&target_path_manifest)
+            .arg(&target_path_oci)
+            .arg("--remove-signatures")
+            .status()?;
+
+        if !status.success() {
+            return Err(anyhow!(format!("failed to copy image: {:?}", status)));
+        }
+
+        // To save space delete the manifest.
+        // TODO LATER - when verify image is added, this will need moving the end of that, if required
+        fs::remove_dir_all(&manifest_path)?;
+        Ok(())
+    }
+
+    fn unpack_image(cid: &str) -> Result<()> {
+        let source_path_oci = format!("/tmp/{}/image_oci:latest", cid);
+        let target_path_bundle = format!("{}{}{}", CONTAINER_BASE, "/", cid);
+
+        info!(sl!(), "cid is {:?}", cid);
+        info!(sl!(), "target_path_bundle is {:?}", target_path_bundle);
+
+        // Unpack image
+
+        let status: ExitStatus = Command::new(UMOCI_PATH)
+            .arg("--verbose")
+            .arg("unpack")
+            .arg("--image")
+            .arg(&source_path_oci)
+            .arg(&target_path_bundle)
+            .status()?;
+
+        if !status.success() {
+            return Err(anyhow!(format!("failed to unpack image: {:?}", status)));
+        }
+
+        // To save space delete the oci image after unpack
+        fs::remove_dir_all(&source_path_oci)?;
+
+        Ok(())
+    }
+
+    async fn pull_image(&self, req: &image::PullImageRequest) -> Result<String> {
+        let image = req.get_image();
+        let mut cid = req.get_container_id();
+
+        if cid.len() == 0 {
+            let v: Vec<&str> = image.rsplit('/').collect();
+            if v[0].len() != 0 {
+                cid = v[0]
+            }
+        }
+
+        let source_creds = (!req.get_source_creds().is_empty()).then(|| req.get_source_creds());
+
+        Self::pull_image_from_registry(image, cid, &source_creds)?;
+        Self::unpack_image(cid)?;
+
+        let mut sandbox = self.sandbox.lock().await;
+        sandbox.images.insert(String::from(image), cid.to_string());
```

Comment:
`images` is written but not read, is it for future use?

Reply:
> This field is just to uniquely represent an image, it will not be processed at runtime, it will be returned to k8s as is, k8s will use this value to call CreateContainer afterwards, then the value will be passed to agent again, this time we can use the value to hash index the Image. PullImage: https://github.com/kubernetes/kubernetes/blob/a2960dc2775848f3b54fb288650e02b4f691743b/pkg/kubelet/images/image_manager.go#L89 CreateContainer https://github.com/kubernetes/kubernetes/blob/a2960dc2775848f3b54fb288650e02b4f691743b/pkg/kubelet/kuberuntime/kuberuntime_container.go#L177

As above,
We need to use `ImageRef` to find the Image bundle directory when you CreateContainer later. Currently, ImageName is used as ImageRef. I think a format like `docker.io/library/busybox:latest` is enough to ensure the uniqueness of Image.
The current HashMap should be considered the simplest ImageManager.

Reply:
Thanks.
Yes, k8s reference the image by `ImageRef` that is returned by `PullImage`. So we save `ImageRef` as key to HashMap. The value of HashMap is `rootfs` (bundle path).

Reply:
Ok, so we should keep it for now, thanks.